File: backend/components/metiers/threed/tests_centralises/services/core/controllers/test_controllers.js
// Test avancé pour controllers.js du module services/core/controllers
// Remplacer l'import par le chemin réel si besoin
// const controllers = require('../../../../services/core/controllers/controllers');
const ServicesController = require('../../../../services/core/controllers/services_controller');

describe('Services Core Controllers', () => {
  it('should expose expected API', () => {
    // expect(controllers).toBeDefined();
    expect(true).toBe(true);
  });

  it('initialise le contrôleur et audite', () => {
    const ctrl = new ServicesController({ mode: 'prod' });
    expect(ctrl.init({ version: 1 })).toBe(true);
    expect(ctrl.config).toEqual({ version: 1 });
    expect(ctrl.getAuditTrail().length).toBe(1);
  });

  it('handle exécute une action métier et audite', () => {
    const ctrl = new ServicesController();
    ctrl.init({});
    const result = ctrl.handle('action', { foo: 'bar' });
    expect(result.success).toBe(true);
    expect(result.action).toBe('action');
    expect(result.payload).toEqual({ foo: 'bar' });
    expect(ctrl.getAuditTrail().pop().action).toBe('handle');
  });

  it('handle lève une erreur sur action invalide', () => {
    const ctrl = new ServicesController();
    ctrl.init({});
    expect(() => ctrl.handle('', {})).toThrow('Invalid action');
    expect(ctrl.getAuditTrail().pop().action).toBe('error');
  });
});

File: backend/components/metiers/threed/tests_centralises/services/core/controllers/test_controllers.py
import types

# Test avancé pour controllers.py du module services/core/controllers
# from components.metiers.threed.services.core.controllers
# import ...


def test_services_core_controllers():
    # Ajouter des assertions avancées selon la logique métier
    assert True


def test_services_controller_nominal():
    from backend.components.metiers.threed.services.core.controllers.services_controller import ServicesController

    ctrl = ServicesController({"mode": "prod"})
    assert ctrl.init({"version": 1}) is True
    result = ctrl.handle("action", {"foo": "bar"})
    assert result["success"] is True
    assert result["action"] == "action"
    assert result["payload"] == {"foo": "bar"}
    assert isinstance(ctrl.get_audit_trail(), list)
    assert ctrl.get_audit_trail()[-1]["action"] == "handle"


def test_services_controller_edge_case():
    from backend.components.metiers.threed.services.core.controllers.services_controller import ServicesController

    ctrl = ServicesController()
    ctrl.init({})
    try:
        ctrl.handle("", {})
    except ValueError as e:
        assert "Invalid action" in str(e)
    assert ctrl.get_audit_trail()[-1]["action"] == "error"


# Mocks pour accès et modèles
class DummyUser:
    def __init__(self, role):
        self.role = role


def test_services_controller_get_model_access():
    from backend.components.metiers.threed.services.core.controllers.services_controller import ServicesController

    ctrl = ServicesController()
    ctrl.audit_log = []
    # Mock check_access et get_threed_model
    ctrl.get_model = types.MethodType(
        lambda self, model_id, user: {"id": model_id, "name": "Test"}, ctrl
    )
    model = ctrl.get_model("m1", DummyUser("admin"))
    assert model["id"] == "m1"


def test_services_controller_list_models_access():
    from backend.components.metiers.threed.services.core.controllers.services_controller import ServicesController

    ctrl = ServicesController()
    ctrl.audit_log = []
    ctrl.list_models = types.MethodType(
        lambda self, user: [{"id": "m1"}, {"id": "m2"}], ctrl
    )
    models = ctrl.list_models(DummyUser("admin"))
    assert isinstance(models, list)
    assert len(models) == 2
